### ISSUE TITLE
Can't install mypy 0.902

### ISSUE THREAD

**Issue by zupo** [nix-community/poetry2nix#323]
I get the following error when trying to install mypy 0.902:

```
Processing /private/tmp/nix-build-python3.8-mypy-0.902.drv-0/mypy-0.902
    ERROR: Command errored out with exit status 1:
     command: /nix/store/3dvkfi06mzgszfwkhlrjnww1fj02ymbf-python3-3.8.9/bin/python3.8 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/tmp/nix-build-python3.8-mypy-0.902.drv-0/pip-req-build-8zacv0o5/setup.py'"'"'; __file__='"'"'/private/tmp/nix-build-python3.8-mypy-0.902.drv-0/pip-req-build-8zacv0o5/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /private/tmp/nix-build-python3.8-mypy-0.902.drv-0/pip-pip-egg-info-s0q3yqyp
         cwd: /private/tmp/nix-build-python3.8-mypy-0.902.drv-0/pip-req-build-8zacv0o5/
    Complete output (84 lines):
    mypy/modulefinder.py:432: error: Library stubs not installed for "toml" (or incompatible with Python 3.8)
    mypy/config_parser.py:10: error: Library stubs not installed for "toml" (or incompatible with Python 3.8)
    mypy/config_parser.py:10: note: Hint: "python3 -m pip install types-toml"
    mypy/config_parser.py:10: note: (or run "mypy --install-types" to install all missing stub packages)
    mypy/config_parser.py:10: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
    mypy/fastparse.py:109: error: Library stubs not installed for "typed_ast" (or incompatible with Python 3.8)
    mypy/fastparse.py:109: note: Hint: "python3 -m pip install types-typed-ast"
    mypy/fastparse2.py:57: error: Library stubs not installed for "typed_ast.ast27" (or incompatible with Python 3.8)
    mypy/fastparse2.py:68: error: Library stubs not installed for "typed_ast" (or incompatible with Python 3.8)
...
```

Because I saw `Hint: "python3 -m pip install types-toml"` in the error log, I tried to work around it with the following:

```
    mypy = super.mypy.overridePythonAttrs (old: {
      buildInputs = (old.buildInputs or []) ++ [
        self.types-toml
      ];
    });
```

But this gives me another error:

```
error: attribute 'types-toml' missing
```

What else can I try?

**Comment by NorfairKing** [comment]
@zupo I didn't see the `types-toml` problem, but I managed to get your override to work by adding `types-toml` in `pyproject.toml` and running `poetry lock --no-update`.

**Comment by zupo** [comment]
> @zupo I didn't see the `types-toml` problem, but I managed to get your override to work by adding `types-toml` in `pyproject.toml` and running `poetry lock --no-update`.

Ah that is a good point!

**Comment by Shados** [comment]
I'm seeing the same failure, with mypy 0.910. Adding types-toml to pyproject/poetry.lock and adding it to `buildInputs` in the poetry overrides does *not* change the error.

**Comment by Shados** [comment]
Update: It looks like #296 is the cause. If I disable that like so:
```
      mypy = super.mypy.overridePythonAttrs (oa: {
        MYPY_USE_MYPYC = false;
      });
```
mypy builds fine.

**Comment by zupo** [comment]
Note that `MYPY_USE_MYPYC` makes mypy run 5x to 10x faster, so it'd be great if we can keep it.

**Comment by adisbladis** [comment]
I think this issue can be fixed by adding an override that creates a `py.typed` in the `toml` build.

**Comment by NorfairKing** [comment]
@adisbladis could you elaborate? Why does that work? How would you do that?

**Comment by takeda** [comment]
@Shados there's undocumented option in poetry2nix called `preferWheels` it could be used as a workaround for now (as these dependencies are only needed for compiling mypy). Also just above my response there is a PR in progress for mypy 0.910 it looks like author correctly identified what was needed (I had problem figuring out what was missing to compile it).

I'm wondering if `poetry2nix` can reference `types-toml` and `types-typed-ast` from nixpkgs when compiling (would need to wait for that PR to merge), will require to add it to `pyproject.toml` (hope not, but maybe as a workaround for now) or perhaps will copy these definitions to the `overrides.nix`?

**Comment by l0b0** [comment]
tl;dr [Disabling mypyc](https://github.com/nix-community/poetry2nix/issues/323#issuecomment-871929102) seems to be the only thing which makes mypy 0.910 build.

The original shell.nix:

```nix
let
  pkgs = import <nixpkgs> { };
  poetryEnv = pkgs.poetry2nix.mkPoetryEnv {
    python = pkgs.python39;
    projectDir = ./.;

    # Work around <https://github.com/nix-community/poetry2nix/issues/322>
    overrides = [
      pkgs.poetry2nix.defaultPoetryOverrides
      (self: super: {
        pytest-randomly = super.pytest-randomly.overrideAttrs (old: {
          postPatch = old.postPatch or ''sed -i 's/importlib-metadata >= 3.6.0 ; python_version < "3.10"//' setup.cfg'';
        });
        ruamel-yaml = super.ruamel-yaml.overridePythonAttrs (old: {
          propagatedBuildInputs = (old.propagatedBuildInputs or [ ]) ++ [ self.ruamel-yaml-clib ];
        });
      })
    ];
  };
in
pkgs.mkShell {
  buildInputs = [
    poetryEnv
  ];
}
```

Trying to build this in a repository which requires mypy 0.910 ran into #322:

```text
Processing /tmp/nix-build-python3.9-mypy-0.910.drv-0/mypy-0.910
    ERROR: Command errored out with exit status 1:
     command: /nix/store/nki9ywqzbvz68vr75kn2r7g1q84f5agy-python3-3.9.6/bin/python3.9 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/nix-build-python3.9-mypy-0.910.drv-0/pip-req-build-t7zxapmj/setup.py'"'"'; __file__='"'"'/tmp/nix-build-python3.9-mypy-0.910.drv-0/pip-req-build-t7zxapmj/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/nix-build-python3.9-mypy-0.910.drv-0/pip-pip-egg-info-qjoki4uq
         cwd: /tmp/nix-build-python3.9-mypy-0.910.drv-0/pip-req-build-t7zxapmj/
    Complete output (84 lines):
    mypy/modulefinder.py:436: error: Library stubs not installed for "toml" (or incompatible with Python 3.9)
    mypy/config_parser.py:10: error: Library stubs not installed for "toml" (or incompatible with Python 3.9)
    mypy/config_parser.py:10: note: Hint: "python3 -m pip install types-toml"
    mypy/config_parser.py:10: note: (or run "mypy --install-types" to install all missing stub packages)
    mypy/config_parser.py:10: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
    mypy/fastparse.py:109: error: Library stubs not installed for "typed_ast" (or incompatible with Python 3.9)
    mypy/fastparse.py:109: note: Hint: "python3 -m pip install types-typed-ast"
    mypy/fastparse2.py:57: error: Library stubs not installed for "typed_ast.ast27" (or incompatible with Python 3.9)
    mypy/fastparse2.py:68: error: Library stubs not installed for "typed_ast" (or incompatible with Python 3.9)
    mypy/fastparse2.py:133: error: Argument 1 to "is_no_type_check_decorator" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:173: error: Type of variable becomes "Dict[type, Callable[[Optional[Any]], Any]]" due to an unfollowed import
    mypy/fastparse2.py:181: error: Argument 2 to "visit" becomes "Optional[Any]" due to an unfollowed import
    mypy/fastparse2.py:192: error: Argument 3 to "set_line" becomes "Union[Any, Any, Any]" due to an unfollowed import
    mypy/fastparse2.py:197: error: Argument 2 to "translate_expr_list" becomes "Sequence[Any]" due to an unfollowed import
    mypy/fastparse2.py:205: error: Argument 2 to "get_lineno" becomes "Union[Any, Any]" due to an unfollowed import
    mypy/fastparse2.py:210: error: Argument 2 to "translate_stmt_list" becomes "Sequence[Any]" due to an unfollowed import
    mypy/fastparse2.py:229: error: Argument 2 to "translate_type_comment" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:244: error: Type of variable becomes "Dict[Type[Any], str]" due to an unfollowed import
    mypy/fastparse2.py:259: error: Argument 2 to "from_operator" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:268: error: Type of variable becomes "Dict[Type[Any], str]" due to an unfollowed import
    mypy/fastparse2.py:281: error: Argument 2 to "from_comp_operator" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:288: error: Argument 2 to "as_block" becomes "List[Any]" due to an unfollowed import
    mypy/fastparse2.py:295: error: Argument 2 to "as_required_block" becomes "List[Any]" due to an unfollowed import
    mypy/fastparse2.py:346: error: Argument 2 to "visit_Module" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:366: error: Argument 2 to "visit_FunctionDef" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:479: error: Argument 2 to "transform_args" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:528: error: Argument 2 to "extract_names" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:536: error: Argument 3 to "convert_arg" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:570: error: Argument 2 to "stringify_name" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:583: error: Argument 2 to "visit_ClassDef" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:599: error: Argument 2 to "visit_Return" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:604: error: Argument 2 to "visit_Delete" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:614: error: Argument 2 to "visit_Assign" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:622: error: Argument 2 to "visit_AugAssign" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:629: error: Argument 2 to "visit_For" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:639: error: Argument 2 to "visit_While" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:646: error: Argument 2 to "visit_If" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:653: error: Argument 2 to "visit_With" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:661: error: Argument 2 to "visit_Raise" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:677: error: Argument 2 to "visit_TryExcept" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:681: error: Argument 2 to "visit_TryFinally" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:688: error: Argument 2 to "try_handler" becomes "List[Any]" due to an unfollowed import
    mypy/fastparse2.py:688: error: Argument 3 to "try_handler" becomes "List[Any]" due to an unfollowed import
    mypy/fastparse2.py:688: error: Argument 4 to "try_handler" becomes "List[Any]" due to an unfollowed import
    mypy/fastparse2.py:688: error: Argument 5 to "try_handler" becomes "List[Any]" due to an unfollowed import
    mypy/fastparse2.py:714: error: Argument 2 to "visit_Print" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:718: error: Argument 2 to "visit_Exec" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:724: error: Argument 2 to "visit_Repr" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:729: error: Argument 2 to "visit_Assert" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:734: error: Argument 2 to "visit_Import" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:750: error: Argument 2 to "visit_ImportFrom" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:766: error: Argument 2 to "visit_Global" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:771: error: Argument 2 to "visit_Expr" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:777: error: Argument 2 to "visit_Pass" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:782: error: Argument 2 to "visit_Break" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:787: error: Argument 2 to "visit_Continue" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:794: error: Argument 2 to "visit_BoolOp" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:815: error: Argument 2 to "visit_BinOp" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:825: error: Argument 2 to "visit_UnaryOp" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:843: error: Argument 2 to "visit_Lambda" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:858: error: Argument 2 to "visit_IfExp" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:865: error: Argument 2 to "visit_Dict" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:871: error: Argument 2 to "visit_Set" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:876: error: Argument 2 to "visit_ListComp" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:877: error: Target type of cast becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:881: error: Argument 2 to "visit_SetComp" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:882: error: Target type of cast becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:886: error: Argument 2 to "visit_DictComp" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:899: error: Argument 2 to "visit_GeneratorExp" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:911: error: Argument 2 to "visit_Yield" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:916: error: Argument 2 to "visit_Compare" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:924: error: Argument 2 to "visit_Call" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:925: error: Type of variable becomes "List[Any]" due to an unfollowed import
    mypy/fastparse2.py:956: error: Argument 2 to "visit_Num" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:982: error: Argument 2 to "visit_Str" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:1000: error: Argument 2 to "visit_Ellipsis" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:1004: error: Argument 2 to "visit_Attribute" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:1020: error: Argument 2 to "visit_Subscript" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:1030: error: Argument 2 to "visit_Name" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:1035: error: Argument 2 to "visit_List" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:1045: error: Argument 2 to "visit_Tuple" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:1052: error: Argument 2 to "visit_Slice" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:1058: error: Argument 2 to "visit_ExtSlice" becomes "Any" due to an unfollowed import
    mypy/fastparse2.py:1062: error: Argument 2 to "visit_Index" becomes "Any" due to an unfollowed import
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

[Installing types-toml](https://github.com/nix-community/poetry2nix/issues/323#issuecomment-862983226) and types-typed-ast does not help. The next suggestion was [`preferWheels = true;`](https://github.com/nix-community/poetry2nix/issues/323#issuecomment-882736451), which resulted in the following:

```console
$ nix-shell 
these derivations will be built:
  /nix/store/g8dgvfml8bdznbzfvp0ys8rjdlycmvas-python3.9-tomlkit-0.7.2.drv
  /nix/store/ci9aiykp4czz57ra0yjbjmz7cj47gbzg-python3.9-toml-sort-0.19.0.drv
  /nix/store/h25yiif65ggm1xc2w6z5q7h6kbpmgial-python3.9-urllib3-1.26.6.drv
  /nix/store/nqgjnfjq303ddl2xddvw4cxqc07sxydx-python3.9-requests-2.25.1.drv
  /nix/store/xcd6r4zz4yl33dzfjn6h9jqkf4xnj8xi-python3.9-ruamel-yaml-0.17.10.drv
  /nix/store/02k6klk7jbvrmr7f8mzwq682r4lcsycp-python3.9-language-formatters-pre-commit-hooks-2.1.0.drv
  /nix/store/0g350rp4ql0vr6ghj8pr50cfrn3b2qic-python3.9-typing-extensions-3.7.4.3.drv
  /nix/store/izmpaymzgh7xcjscm9493czsym0aqq7r-python3.9-pytest-6.2.4.drv
  /nix/store/600fnk9bkaqxqb4cd2mnylj8fclmmn23-python3.9-pytest-runner-5.3.1.drv
  /nix/store/bk5v7v3wr0zyq7lxz5q68skd9f5hakqv-python3.9-wrapt-1.12.1.drv
  /nix/store/1wq8qsaznnyy3ahx3xb8zd6mcd8196d1-python3.9-astroid-2.6.5.drv
  /nix/store/6n7cxaqapgpmcyhx5smfqfjg1f2j0ldc-python3.9-virtualenv-20.1.0.drv
  /nix/store/8khrvy1s63f65acqar53spimajg8jl0n-python3.9-mccabe-0.6.1.drv
  /nix/store/hsw4l88mlypx33kfq7q20vk0b0hb9xb0-python3.9-pre-commit-2.13.0.drv
  /nix/store/m1gxhylcc2fi2ybavw3a0sghidadmija-python3.9-mypy-0.910.drv
  /nix/store/v65npj45av86gk51sjqpwxwxf2ykh7fa-python3.9-pylint-2.9.5.drv
  /nix/store/y5fq55vlhlrbl42ndiplaxfd31nw2r1m-python3.9-pytest-randomly-3.10.1.drv
  /nix/store/7g0slp6ybmvlbymylgw67kf5xsdvg8hz-python3-3.9.6-env.drv
building '/nix/store/izmpaymzgh7xcjscm9493czsym0aqq7r-python3.9-pytest-6.2.4.drv'...
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing wheel setup hook
Using wheelUnpackPhase
Sourcing pip-install-hook
Using pipInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
unpacking sources
Executing wheelUnpackPhase
Finished executing wheelUnpackPhase
patching sources
sed: can't read setup.cfg: No such file or directory
builder for '/nix/store/izmpaymzgh7xcjscm9493czsym0aqq7r-python3.9-pytest-6.2.4.drv' failed with exit code 2
cannot build derivation '/nix/store/7g0slp6ybmvlbymylgw67kf5xsdvg8hz-python3-3.9.6-env.drv': 1 dependencies couldn't be built
error: build of '/nix/store/7g0slp6ybmvlbymylgw67kf5xsdvg8hz-python3-3.9.6-env.drv' failed
```
